packages/experience-builder-sdk/src/blocks/VisualEditorBlock.tsx

```diff
@@ -40,18 +45,37 @@ type VisualEditorBlockProps = {
   areEntitiesFetched: boolean;
 };
 
+const DESIGN_COMPONENT_NODE_TYPES = ['designComponent', 'designComponentBlock'];
+
 export const VisualEditorBlock = ({
-  node,
+  node: rawNode,
   dataSource,
   unboundValues,
   resolveDesignValue,
   entityStore,
   areEntitiesFetched,
 }: VisualEditorBlockProps) => {
-  const componentRegistration = useMemo(
-    () => getComponentRegistration(node.data.blockId as string),
-    [node]
-  );
+  const node = useMemo(() => {
+    if (rawNode.type === 'designComponent' && areEntitiesFetched) {
+      return resolveDesignComponent({
+        node: rawNode,
+        entityStore: entityStore.current,
+      });
+    }
+
+    return rawNode;
+  }, [areEntitiesFetched, entityStore, rawNode]);
+
+  const componentRegistration = useMemo(() => {
+    const registration = getComponentRegistration(node.data.blockId as string);
+    if (node.type === 'designComponent' && !registration) {
+      return createDesignComponentRegistration({
+        definitionId: node.data.blockId as string,
+        component: DesignComponent,
+      });
+    }
+    return registration;
+  }, [node]);
```

Comment:
Theoretically, this would have no effect on the performance of this component if this is your concern. tbh I kinda like that one memo returns one variable and thus split up as we need both variable below.

Reply:
+1. As I said above, I like how each useMemo returns one variable and have separate concerns.